### PR TITLE
Refactor the Wine Engine Object

### PR DESCRIPTION
**Pull request by madoar**
This PR refactors the Wine engine object with newer JS language features like `class` and streams

### CODE REVIEW

File: Engines/Wine/Engine/Object/script.js [hunk 1]
```diff
@@ -5,333 +5,362 @@ include("utils.functions.net.download");
...
-            return "";
+                const distribution = containerConfiguration.readValue("wineDistribution", "upstream");
+                const architecture = containerConfiguration.readValue("wineArchitecture", "x86");
+                
+                const operatingSystem = this.operatingSystemFetcher.fetchCurrentOperationSystem().getWinePackage();
+
+                subCategory = distribution + "-" + operatingSystem + "-" + architecture;
+                version = containerConfiguration.readValue("wineVersion");
+            } else {
+                print('Wine prefix "' + this.prefixDirectory() + '" does not exist!');
```
Comment (madoar) [inline Engines/Wine/Engine/Object/script.js]: Should we throw an `Error` here?

File: Engines/Wine/Engine/Object/script.js [hunk 2]
```diff
@@ -5,333 +5,362 @@ include("utils.functions.net.download");
...
-};
+    /**
+     * returns the path to the engine binary directory
+     * if no parameters are given, the Wine version of the current prefix is used
+     * @param {string} [subCategory] Wine sub-category
+     * @param {string} [version] Wine version
+     * @returns {string} path to "wine" binary
+     */
+    binPath(subCategory, version) {
+        if (0 == arguments.length) {
```
Comment (madoar) [inline Engines/Wine/Engine/Object/script.js]: I don't like the `arguments.length == x` checks. Maybe we can find a better solution to do this?

File: Engines/Wine/Engine/Object/script.js [hunk 3]
```diff
@@ -5,333 +5,362 @@ include("utils.functions.net.download");
...
+        const appEscaped = application.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
+        const re = new RegExp("(.*)\\|\\|\\|.*" + appEscaped);
+        const uuid = list.match(re);
+        if (uuid) {
+            this._implementation.getWizard().wait(tr("Please wait while {0} is uninstalled...", application));
+            this.run("uninstaller", ["--remove", uuid[1]], this.prefixDirectory(), false, true);
+
+            return true;
+        } else {
+            print(tr("Could not uninstall {0}!", application));
```
Comment (madoar) [inline Engines/Wine/Engine/Object/script.js]: Should we throw an `Error` here?

**Comment by madoar** [comment]
Additionally we should really add some more comments to this class!

File: Engines/Wine/Engine/Object/script.js [hunk 4]
```diff
@@ -5,333 +5,362 @@ include("utils.functions.net.download");
...
+     *
+     * @returns {string} architecture ("x86" or "amd64")
+     */
+    architecture() {
         if (fileExists(this.prefixDirectory())) {
-            var configFactory = Bean("compatibleConfigFileFormatFactory");
-            var containerConfiguration = configFactory.open(this.prefixDirectory() + "/phoenicis.cfg");
-            var architecture = containerConfiguration.readValue("wineArchitecture", "x86");
+            const containerConfiguration = this.configFactory.open(this.prefixDirectory() + "/phoenicis.cfg");
+            const architecture = containerConfiguration.readValue("wineArchitecture", "x86");
```
[the remark this answers is not in the document]
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: I'm not sure that a json is clearer compared to `readValue`. I would prefer 3 methods (`getContainerArchitecture` etc.).

File: Engines/Wine/Engine/Object/script.js [hunk 5]
```diff
@@ -5,333 +5,362 @@ include("utils.functions.net.download");
...
-var LATEST_DEVELOPMENT_VERSION = "4.11";
+const LATEST_DEVELOPMENT_VERSION = "4.11";
 /* exported LATEST_STAGING_VERSION */
-var LATEST_STAGING_VERSION = "4.11";
+const LATEST_STAGING_VERSION = "4.11";
 /* exported LATEST_DOS_SUPPORT_VERSION */
-var LATEST_DOS_SUPPORT_VERSION = "4.0";
+const LATEST_DOS_SUPPORT_VERSION = "4.0";
+
+const FilenameUtils = Java.type("org.apache.commons.io.FilenameUtils");
```
[the remark this answers is not in the document]
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: @qparis should state his opinion about this as well but for me your proposal makes sense.

We should also consider to do something like:
```Javascript
const includedClass = include("...");
```

File: Engines/Wine/Engine/Object/script.js [hunk 6]
```diff
@@ -5,333 +5,362 @@ include("utils.functions.net.download");
...
-            var architecture = containerConfiguration.readValue("wineArchitecture", "x86");
+            const containerConfiguration = this.configFactory.open(this.prefixDirectory() + "/phoenicis.cfg");
+            const architecture = containerConfiguration.readValue("wineArchitecture", "x86");
+
             return architecture;
-        }
-        else {
-            print("Wine prefix \"" + this.prefixDirectory() + "\" does not exist!");
+        } else {
+            print('Wine prefix "' + this.prefixDirectory() + '" does not exist!');
```
[the remark this answers is not in the document]
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: I'm not sure. Is it valid to call the method if there is no prefix?

File: Engines/Wine/Engine/Object/script.js (hunk 1 above)
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: Yes.

File: Engines/Wine/Engine/Object/script.js (hunk 2 above)
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: You can check if the parameter is defined if you prefer that.

File: Engines/Wine/Engine/Object/script.js (hunk 3 above)
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: No. It is valid to try to uninstall an app which does not exist. Returning `false` should be sufficient.

File: Engines/Wine/Engine/Object/script.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: The main issue for me is the extraction code and the access to `containerConfiguration` which I would like to hide behind a method.

In addition it is quite easy to access the values:

```javascript
const { architecture, distribution, version } = fetchContainerConfiguration(this.prefixDirectory() + "/phoenicis.cfg");
```

File: Engines/Wine/Engine/Object/script.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: I don't know. The printed message suggests to me that it is not valid

File: Engines/Wine/Engine/Object/script.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: In my opinion, the `architecture` method already encapsulates this.

File: Engines/Wine/Engine/Object/script.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: We can try the error and check if app installations etc still work.

File: Engines/Wine/Engine/Object/script.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: I don't understand? The `architecture()` method only returns a single value, while I suggest to return three wine related values at once.

File: Engines/Wine/Engine/Object/script.js (hunk 3 above)
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: Ok

File: Engines/Wine/Engine/Object/script.js (hunk 1 above)
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: Done

File: Engines/Wine/Engine/Object/script.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: Done (error message added)

File: Engines/Wine/Engine/Object/script.js (hunk 2 above)
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: Let us retain it as it is for now

File: Engines/Wine/Engine/Object/script.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: I know. However, your point is that you want to hide something in a method. In my opinion the `architecture()` method already does so. Therefore I do not see the need for additional encapsulation.

[the remark this answers is not in the document]
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: Let me demonstrate what I want to do in a different (new) PR. This is out of scope anyway.

File: Engines/Wine/Engine/Object/script.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: I will create a new issue to track this.

File: Engines/Wine/Engine/Object/script.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (plata) [reply Engines/Wine/Engine/Object/script.js]: Ok.

**Comment by madoar** [comment]
About the Codacy error: I don't want to fix the issue because it seems suspect to me

File: Engines/Wine/Engine/Object/script.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (madoar) [reply Engines/Wine/Engine/Object/script.js]: See #1070

**Comment by plata** [comment]
You're right about the Codacy error.

**Pull request merged**